pkg/controller/perconaservermongodb/finalizers.go

```diff
@@ -55,44 +56,106 @@ func (r *ReconcilePerconaServerMongoDB) checkFinalizers(ctx context.Context, cr
 }
 
 func (r *ReconcilePerconaServerMongoDB) deletePSMDBPods(ctx context.Context, cr *api.PerconaServerMongoDB) (err error) {
-	done := true
-	for _, rs := range cr.Spec.Replsets {
-		sts, err := r.getRsStatefulset(ctx, cr, rs.Name)
-		if err != nil {
-			if k8serrors.IsNotFound(err) {
-				continue
+	if cr.Spec.Sharding.Enabled {
+		cr.Spec.Sharding.Mongos.Size = 0
+
+		sts := new(appsv1.StatefulSet)
+		err := r.client.Get(ctx, cr.MongosNamespacedName(), sts)
+		if client.IgnoreNotFound(err) != nil {
+			return errors.Wrap(err, "failed to get mongos statefulset")
+		}
+		if sts.Spec.Replicas != nil && *sts.Spec.Replicas > 0 {
+			err = r.disableBalancer(ctx, cr)
+			if err != nil {
+				return errors.Wrap(err, "failed to disable balancer")
 			}
-			return errors.Wrap(err, "get rs statefulset")
 		}
-
-		pods := &corev1.PodList{}
-		err = r.client.List(ctx,
-			pods,
-			&client.ListOptions{
-				Namespace:     cr.Namespace,
-				LabelSelector: labels.SelectorFromSet(sts.Spec.Selector.MatchLabels),
-			},
-		)
+		list, err := r.getMongosPods(ctx, cr)
 		if err != nil {
-			if k8serrors.IsNotFound(err) {
-				continue
-			}
-			return errors.Wrap(err, "get rs statefulset")
+			return errors.Wrap(err, "get mongos pods")
 		}
-		if len(pods.Items) > int(*sts.Spec.Replicas) {
+		if len(list.Items) != 0 {
 			return errWaitingTermination
 		}
-		if *sts.Spec.Replicas != 1 {
-			rs.Size = 1
-			done = false
+	}
+
+	replsetsDeleted := true
+	for _, rs := range cr.Spec.Replsets {
+		if err := r.deleteRSPods(ctx, cr, rs); err != nil {
+			if err == errWaitingTermination {
+				replsetsDeleted = false
+				continue
+			}
+			return err
 		}
 	}
-	if !done {
+	if !replsetsDeleted {
 		return errWaitingTermination
 	}
+
+	if cr.Spec.Sharding.Enabled && cr.Spec.Sharding.ConfigsvrReplSet != nil {
+		if err := r.deleteRSPods(ctx, cr, cr.Spec.Sharding.ConfigsvrReplSet); err != nil {
+			return err
+		}
+	}
```

Comment:
rather than handling cfg replset separately, can we append it to the replsets list? i.e https://github.com/percona/percona-server-mongodb-operator/blob/main/pkg/controller/perconaservermongodb/psmdb_controller.go#L289-L292 (if cfg rs doesn't need to be first, it can be simplified)

this way we'll also have termination error handling for cfg too

Reply:
cfg replset should be terminated in the end after all other replsets are stopped: https://www.mongodb.com/docs/manual/tutorial/restart-sharded-cluster/#stop-sharded-cluster

In the loop, we resize all other replsets at once and wait until they are all stopped. Only after this, we terminate the cfg replset.

If we want to add it to replset list and keep the same logic, it will look like this:
```
func (r *ReconcilePerconaServerMongoDB) deletePSMDBPods(ctx context.Context, cr *api.PerconaServerMongoDB) (err error) {
	var repls []*api.ReplsetSpec
	repls = append(repls, cr.Spec.Replsets...)

	if cr.Spec.Sharding.Enabled {
		cr.Spec.Sharding.Mongos.Size = 0
		...
		if len(list.Items) != 0 {
			return errWaitingTermination
		}
		if cr.Spec.Sharding.ConfigsvrReplSet != nil {
			repls = append(repls, cr.Spec.Sharding.ConfigsvrReplSet)
		}
	}

	replsetsDeleted := true
	for _, rs := range repls {
		if rs.Name == api.ConfigReplSetName && !replsetsDeleted {
			break
		}
		if err := r.deleteRSPods(ctx, cr, rs); err != nil {
			if err == errWaitingTermination {
				replsetsDeleted = false
				continue
			}
			return err
		}
	}
	if !replsetsDeleted {
		return errWaitingTermination
	}
	return nil
}
```

I think the current implementation is more clear and simple than in this example